File: solutions/0049-group-anagrams/group-anagrams.cpp
// Given an array of strings, group anagrams together.
//
// Example:
//
//
// Input: ["eat", "tea", "tan", "ate", "nat", "bat"],
// Output:
// [
//   ["ate","eat","tea"],
//   ["nat","tan"],
//   ["bat"]
// ]
//
// Note:
//
//
// 	All inputs will be in lowercase.
// 	The order of your output does not matter.
//
//


class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> mp;
        for(string s : strs) {
            string key = s; 
            sort(key.begin(), key.end());
            mp[key].push_back(s);
        }
        
        vector<vector<string>> anagrams;
        for(auto p : mp) { 
            anagrams.push_back(p.second);
        }
        return anagrams;
    }
    
    
// public:
//     vector<vector<string>> groupAnagrams(vector<string>& strs) {
//         unordered_map<string, vector<string>> mp;
//         for (string s : strs) {
//             // 由于都是小写字母，可以采用计数排序
//             mp[strSort(s)].push_back(s);
//         }
//         vector<vector<string>> anagrams;
//         for (auto p : mp) { 
//             anagrams.push_back(p.second);
//         }
//         return anagrams;
//     }
// private:
//     string strSort(string s) {
//         int counter[26] = {0};
//         for (char c : s) {
//             counter[c - 'a']++;
//         }
//         string t;
//         for (int c = 0; c < 26; c++) {
//             t += string(counter[c], c + 'a');
//         }
//         return t;
//     }
};

File: solutions/0049-group-anagrams/group-anagrams.java
// Given an array of strings, group anagrams together.
//
// Example:
//
//
// Input: ["eat", "tea", "tan", "ate", "nat", "bat"],
// Output:
// [
//   ["ate","eat","tea"],
//   ["nat","tan"],
//   ["bat"]
// ]
//
// Note:
//
//
// 	All inputs will be in lowercase.
// 	The order of your output does not matter.
//
//


class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        if(strs.length == 0 || strs == null) return new ArrayList<List<String>>();
        
        HashMap<String, ArrayList<String>> map = new HashMap<String, ArrayList<String>>();
        for(int i=0; i<strs.length; ++i) {
            char[] S = strs[i].toCharArray();
            // bubbleSort(S);
            Arrays.sort(S);
            String std = new String(S);
            if(!map.containsKey(std)) map.put(std, new ArrayList<String>());
            // HashMap<key,ArrayList<Values>>
            // map.get(key)获取到的是ArrayList<Values> 这个对象。
            map.get(std).add(strs[i]);
        }
        
        return new ArrayList<List<String>>(map.values());
    }
    
    public void bubbleSort(char[] S) {
        for(int i=0; i<S.length-1; ++i) {
            for(int j=0; j<S.length-i-1; ++j) {
                if(S[j] > S[j+1]) {
                    char tmp = S[j+1];
                    S[j+1] = S[j];
                    S[j] = tmp;
                }
            }
        }
    }
}

